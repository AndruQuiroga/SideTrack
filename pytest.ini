[pytest]
markers =
    unit: unit tests
    integration: integration tests
    contract: contract tests with external APIs
    slow: tests exceeding the default speed budget
    gpu: tests requiring GPU hardware
    e2e: end-to-end tests
addopts = -q -x --strict-markers --disable-warnings -m "not contract"
