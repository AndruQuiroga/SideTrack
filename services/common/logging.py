import logging
import sys

import structlog


def setup_logging(level: int = logging.INFO) -> None:
    """Configure structlog for JSON output."""
    logging.basicConfig(format="%(message)s", stream=sys.stdout, level=level)
    structlog.configure(
        processors=[
            structlog.processors.TimeStamper(fmt="iso"),
            structlog.stdlib.add_log_level,
            structlog.processors.StackInfoRenderer(),
            structlog.processors.format_exc_info,
            structlog.processors.JSONRenderer(),
        ],
        wrapper_class=structlog.make_filtering_bound_logger(level),
        logger_factory=structlog.stdlib.LoggerFactory(),
    )

