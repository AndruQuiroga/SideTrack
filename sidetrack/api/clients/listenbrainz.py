from __future__ import annotations

from collections.abc import AsyncGenerator
from datetime import UTC, date, datetime
from typing import Any

import httpx
from tenacity import retry, retry_if_exception, stop_after_attempt, wait_exponential


def _retryable(exc: Exception) -> bool:
    if isinstance(exc, httpx.HTTPStatusError):
        status = exc.response.status_code
        return status >= 500 or status == 429
    return isinstance(exc, httpx.RequestError)


_retry = retry(
    reraise=True,
    stop=stop_after_attempt(3),
    wait=wait_exponential(multiplier=1, min=1, max=4),
    retry=retry_if_exception(_retryable),
)


class ListenBrainzClient:
    """Client for the ListenBrainz API."""

    base_url = "https://api.listenbrainz.org/1"

    def __init__(self, client: httpx.AsyncClient) -> None:
        self._client = client

    @_retry
    async def _get(self, url: str, **kwargs: Any) -> httpx.Response:
        resp = await self._client.get(url, timeout=30, **kwargs)
        resp.raise_for_status()
        return resp

    async def fetch_listens(
        self,
        user: str,
        since: date | None,
        token: str | None = None,
        limit: int = 500,
    ) -> list[dict[str, Any]]:
        """Fetch listens for ``user`` from ListenBrainz."""
        params: dict[str, Any] = {"count": min(limit, 1000)}
        if since:
            params["min_ts"] = int(
                datetime.combine(since, datetime.min.time(), tzinfo=UTC).timestamp()
            )
        headers = {"Authorization": f"Token {token}"} if token else None
        url = f"{self.base_url}/user/{user}/listens"
        r = await self._get(url, params=params, headers=headers)
        data = r.json()
        return data.get("listens", [])


async def get_listenbrainz_client() -> AsyncGenerator[ListenBrainzClient, None]:
    async with httpx.AsyncClient() as client:
        yield ListenBrainzClient(client)
