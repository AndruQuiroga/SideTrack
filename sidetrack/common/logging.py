"""Logging utilities shared across SideTrack services."""

from __future__ import annotations

import logging
from typing import Union


_DEFAULT_FORMAT = "%(asctime)s %(levelname)s %(name)s: %(message)s"


def _resolve_level(level: Union[int, str, None]) -> int:
    if level is None:
        return logging.INFO
    if isinstance(level, str):
        numeric_level = logging.getLevelName(level.upper())
        if isinstance(numeric_level, int):
            return numeric_level
        raise ValueError(f"Unknown log level: {level}")
    return int(level)


def setup_logging(level: Union[int, str, None] = None, *, force: bool = False) -> None:
    """Configure standard logging with a simple, readable format."""

    resolved_level = _resolve_level(level)
    root_logger = logging.getLogger()
    root_logger.setLevel(resolved_level)

    if root_logger.handlers and not force:
        return

    logging.basicConfig(
        level=resolved_level,
        format=_DEFAULT_FORMAT,
        datefmt="%Y-%m-%d %H:%M:%S",
        force=force,
    )
