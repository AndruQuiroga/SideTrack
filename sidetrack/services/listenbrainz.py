from __future__ import annotations

from collections.abc import AsyncGenerator
from datetime import UTC, date, datetime
from typing import Any

import httpx
import logging
from fastapi import Depends
from tenacity import retry, retry_if_exception, stop_after_attempt, wait_exponential

from sidetrack.api.config import Settings, get_settings

from .base_client import MusicServiceClient
from .models import TrackRef


def _retryable(exc: Exception) -> bool:
    if isinstance(exc, httpx.HTTPStatusError):
        status = exc.response.status_code
        return status >= 500 or status == 429
    return isinstance(exc, httpx.RequestError)


_retry = retry(
    reraise=True,
    stop=stop_after_attempt(3),
    wait=wait_exponential(multiplier=1, min=1, max=4),
    retry=retry_if_exception(_retryable),
)


logger = logging.getLogger(__name__)


class ListenBrainzClient(MusicServiceClient):
    """Client for the ListenBrainz API."""

    source = "listenbrainz"

    base_url = "https://api.listenbrainz.org/1"

    def __init__(self, client: httpx.AsyncClient, *, user: str | None = None, token: str | None = None) -> None:
        self._client = client
        self.user = user
        self.token = token

    @_retry
    async def _get(self, url: str, **kwargs: Any) -> httpx.Response:
        resp = await self._client.get(url, timeout=30, **kwargs)
        resp.raise_for_status()
        return resp

    async def fetch_listens(
        self,
        user: str,
        since: date | None,
        token: str | None = None,
        limit: int = 500,
    ) -> list[dict[str, Any]]:
        """Fetch listens for ``user`` from ListenBrainz."""

        params: dict[str, Any] = {"count": min(limit, 1000)}
        if since:
            params["min_ts"] = int(
                datetime.combine(since, datetime.min.time(), tzinfo=UTC).timestamp()
            )
        headers = {"Authorization": f"Token {token}"} if token else None
        url = f"{self.base_url}/user/{user}/listens"
        r = await self._get(url, params=params, headers=headers)
        data = r.json()
        return data.get("listens", [])

    async def fetch_recently_played(
        self, since: date | None = None, limit: int = 500
    ) -> list[dict[str, Any]]:
        if not self.user:
            raise RuntimeError("ListenBrainz user not configured")
        return await self.fetch_listens(self.user, since, self.token, limit)

    async def get_cf_recommendations(self, user: str, limit: int = 50) -> list[dict[str, Any]]:
        """Fetch collaborative-filtering recommendations for ``user``."""

        url = f"{self.base_url}/user/{user}/cf/recommendations"
        params = {"count": limit}
        resp = await self._get(url, params=params)
        data = resp.json()
        if not isinstance(data, dict):
            logger.warning(
                "unexpected ListenBrainz recommendations payload: %s", data
            )
            raise RuntimeError("unexpected ListenBrainz response format")

        for key in ("recommendations", "recordings", "recommended_recordings"):
            if key in data:
                items = data[key] or []
                if isinstance(items, list):
                    return [i for i in items if isinstance(i, dict)]
                logger.warning(
                    "unexpected ListenBrainz recommendations payload: %s", data
                )
                raise RuntimeError("unexpected ListenBrainz response format")

        logger.warning("unexpected ListenBrainz recommendations payload: %s", data)
        raise RuntimeError("unexpected ListenBrainz response format")

    @staticmethod
    def to_track_ref(rec: dict[str, Any]) -> TrackRef:
        """Convert a ListenBrainz recommendation payload to :class:`TrackRef`."""

        title = rec.get("track_name") or rec.get("title") or ""
        artist = rec.get("artist_name") or rec.get("artist") or ""
        if isinstance(artist, str):
            artists = [artist]
        elif isinstance(artist, list):
            artists = [str(a) for a in artist if a]
        else:
            artists = [str(artist)] if artist else []
        mbid = rec.get("recording_mbid") or rec.get("mbid") or None
        return TrackRef(title=title, artists=artists, lastfm_mbid=mbid)

    async def get_similar_artists(
        self, artist_mbid: str, limit: int = 10
    ) -> list[dict[str, Any]]:
        """Return artists frequently co-listened with ``artist_mbid``."""

        url = f"{self.base_url}/artist/{artist_mbid}/similar"
        params = {"count": limit}
        try:
            resp = await self._get(url, params=params)
        except httpx.HTTPStatusError as exc:
            if exc.response.status_code == 404:
                return []
            raise
        data = resp.json()
        artists = data.get("similar_artists") or data.get("artists") or []
        return [a for a in artists if isinstance(a, dict)]


async def get_listenbrainz_client(
    settings: Settings = Depends(get_settings),
) -> AsyncGenerator[ListenBrainzClient, None]:
    async with httpx.AsyncClient() as client:
        yield ListenBrainzClient(
            client,
            user=settings.listenbrainz_user,
            token=settings.listenbrainz_token,
        )
