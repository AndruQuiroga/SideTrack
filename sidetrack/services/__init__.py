"""Service layer utilities."""

from .listenbrainz import ListenBrainzClient, get_listenbrainz_client
from .listens import ListenService, get_listen_service
from .spotify import SpotifyClient, get_spotify_client
from .lastfm import LastfmClient, get_lastfm_client

__all__ = [
    "ListenService",
    "get_listen_service",
    "SpotifyClient",
    "get_spotify_client",
    "ListenBrainzClient",
    "get_listenbrainz_client",
    "LastfmClient",
    "get_lastfm_client",
]
