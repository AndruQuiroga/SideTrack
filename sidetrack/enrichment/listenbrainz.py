from __future__ import annotations

from typing import List

import httpx

from sidetrack.services.listenbrainz import ListenBrainzClient

from . import TrackRef


class ListenBrainzAdapter:
    """Wrapper around :class:`sidetrack.services.listenbrainz.ListenBrainzClient`."""

    def __init__(self, client: httpx.AsyncClient | None = None) -> None:
        self._httpx = client or httpx.AsyncClient()
        self._owns_client = client is None
        self._client = ListenBrainzClient(self._httpx)

    async def close(self) -> None:
        if self._owns_client:
            await self._httpx.aclose()

    async def get_cf_recommendations(self, user: str, limit: int = 50) -> List[TrackRef]:
        recs = await self._client.get_cf_recommendations(user, limit)
        return [ListenBrainzClient.to_track_ref(r) for r in recs]

    async def get_colisten_similar_artists(
        self, artist_mbid: str, limit: int = 10
    ) -> List[str]:
        """Return artists frequently co-listened with the given artist."""

        artists = await self._client.get_similar_artists(artist_mbid, limit)
        return [a.get("artist_name", "") for a in artists if isinstance(a, dict)]
