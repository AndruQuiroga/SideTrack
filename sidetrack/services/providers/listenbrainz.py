"""ListenBrainz ingestion backend."""

from __future__ import annotations

from typing import Any, List

import httpx

from sidetrack.services.listenbrainz import ListenBrainzClient


class _ProviderListenBrainzClient(ListenBrainzClient):
    """Concrete ListenBrainz client for ingestion purposes."""

    def auth_url(self, callback: str) -> str:  # pragma: no cover - CLI usage
        raise NotImplementedError("ListenBrainz token auth is handled externally")


class ListenBrainzIngester:
    """Fetch listens from the ListenBrainz API."""

    def __init__(self, *, page_size: int = 500) -> None:
        if page_size <= 0:
            raise ValueError("page_size must be positive")
        # ListenBrainz caps count at 1000 per request.
        self.page_size = min(page_size, 1000)

    async def fetch_recent(
        self,
        user: str,
        token: str | None = None,
        since: float | None = None,
    ) -> List[dict[str, Any]]:
        """Return recent listens for ``user``.

        Parameters
        ----------
        user:
            ListenBrainz username.
        token:
            Optional token for authenticated requests.
        since:
            Optional unix timestamp for the earliest listen to return.
        """

        headers = {"Authorization": f"Token {token}"} if token else None
        min_ts = int(since) + 1 if since is not None else None
        listens: list[dict[str, Any]] = []

        async with httpx.AsyncClient() as client:
            lb_client = _ProviderListenBrainzClient(client)
            base_url = f"{lb_client.base_url}/user/{user}/listens"
            url: str | None = base_url
            params: dict[str, Any] | None = {"count": self.page_size}
            sticky_params: dict[str, Any] | None = {"min_ts": min_ts} if min_ts else None

            while url:
                if params and min_ts is not None:
                    params["min_ts"] = min_ts

                resp = await lb_client._get(url, params=params, headers=headers)
                data = resp.json()
                raw_listens, cursor = ListenBrainzClient._extract_listens_payload(data)

                if not raw_listens:
                    break

                oldest_ts: int | None = None
                for raw in raw_listens:
                    if not isinstance(raw, dict):
                        continue
                    raw_ts = raw.get("listened_at")
                    if isinstance(raw_ts, (int, float)):
                        ts = int(raw_ts)
                        if oldest_ts is None or ts < oldest_ts:
                            oldest_ts = ts
                    normalised = self._normalise_listen(raw, user)
                    if normalised is None:
                        continue
                    if since is not None and normalised["listened_at"] <= since:
                        continue
                    listens.append(normalised)

                if since is not None and oldest_ts is not None and oldest_ts <= since:
                    break

                url = ListenBrainzClient._prepare_next_url(
                    base_url,
                    cursor,
                    sticky_params=sticky_params,
                    default_count=self.page_size,
                )
                params = None

        return listens

    @staticmethod
    def _normalise_listen(listen: dict[str, Any], user: str) -> dict[str, Any] | None:
        """Convert a ListenBrainz payload row to a ListenService compatible dict."""

        listened_at = listen.get("listened_at")
        try:
            listened_at_int = int(listened_at)
        except (TypeError, ValueError):
            return None

        raw_metadata = listen.get("track_metadata")
        metadata = dict(raw_metadata) if isinstance(raw_metadata, dict) else {}

        if "track_name" not in metadata and metadata.get("title"):
            metadata["track_name"] = metadata.get("title")
        if "artist_name" not in metadata and metadata.get("artist"):
            metadata["artist_name"] = metadata.get("artist")

        mbid_mapping = metadata.get("mbid_mapping")
        if not isinstance(mbid_mapping, dict):
            mbid_mapping = None

        recording_mbid = None
        if mbid_mapping:
            recording_mbid = mbid_mapping.get("recording_mbid")

        additional_info = metadata.get("additional_info")
        if recording_mbid is None and isinstance(additional_info, dict):
            recording_mbid = additional_info.get("recording_mbid")

        if recording_mbid:
            existing = mbid_mapping or {}
            metadata["mbid_mapping"] = {**existing, "recording_mbid": recording_mbid}
        elif mbid_mapping is not None:
            metadata["mbid_mapping"] = dict(mbid_mapping)

        return {
            "track_metadata": metadata,
            "listened_at": listened_at_int,
            "user_name": listen.get("user_name") or user,
        }
