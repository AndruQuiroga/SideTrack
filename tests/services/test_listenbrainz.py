import logging
from datetime import UTC, date, datetime

import httpx
import pytest

from sidetrack.services.listenbrainz import ListenBrainzClient
from sidetrack.services.recommendation import _listenbrainz_candidates


class _ConcreteListenBrainzClient(ListenBrainzClient):
    def auth_url(self, callback: str) -> str:  # pragma: no cover - not used in tests
        raise NotImplementedError


@pytest.mark.unit
@pytest.mark.asyncio
async def test_listenbrainz_candidates_normalization(snapshot):
    class StubLB:
        async def get_cf_recommendations(self, user: str, limit: int = 50):
            return [
                {
                    "recording_mbid": "rec1",
                    "recording_name": "Track A",
                    "artist_name": "Artist A",
                    "score": 0.9,
                },
                {
                    "recording_mbid": "rec2",
                    "recording_name": "Track B",
                    "artist_name": "Artist B",
                    "score": 0.8,
                },
            ]

    result = await _listenbrainz_candidates(StubLB(), "alice")
    snapshot.assert_match([c.to_dict() for c in result])


@pytest.mark.contract
@pytest.mark.asyncio
@pytest.mark.parametrize("key", ["recommendations", "recordings", "recommended_recordings"])
async def test_get_cf_recommendations_contract(snapshot, respx_mock, key):
    payload = {
        key: [
            {
                "recording_mbid": "rec1",
                "recording_name": "Track A",
                "artist_name": "Artist A",
                "score": 0.9,
            },
            {
                "recording_mbid": "rec2",
                "recording_name": "Track B",
                "artist_name": "Artist B",
                "score": 0.8,
            },
        ]
    }

    respx_mock.get(
        "https://api.listenbrainz.org/1/user/alice/cf/recommendations"
    ).respond(200, json=payload)

    async with httpx.AsyncClient() as client:
        service = _ConcreteListenBrainzClient(client)
        result = await service.get_cf_recommendations("alice")

    snapshot.assert_match(result)


@pytest.mark.contract
@pytest.mark.asyncio
async def test_get_cf_recommendations_unexpected_payload(respx_mock, caplog):
    respx_mock.get(
        "https://api.listenbrainz.org/1/user/alice/cf/recommendations"
    ).respond(200, json={"foo": []})

    caplog.set_level(logging.WARNING)
    async with httpx.AsyncClient() as client:
        service = _ConcreteListenBrainzClient(client)
        with pytest.raises(RuntimeError):
            await service.get_cf_recommendations("alice")

    assert "unexpected ListenBrainz recommendations payload" in caplog.text


@pytest.mark.unit
@pytest.mark.asyncio
async def test_fetch_listens_paginates_until_limit(respx_mock):
    page_one = [
        {"listened_at": 400},
        {"listened_at": 350},
    ]
    page_two = [
        {"listened_at": 300},
        {"listened_at": 250},
    ]

    route = respx_mock.get("https://api.listenbrainz.org/1/user/alice/listens")
    route.side_effect = [
        httpx.Response(
            200,
            json={
                "payload": {
                    "listens": page_one,
                    "next": {"next": "cursor-1", "payload": "payload-1"},
                }
            },
        ),
        httpx.Response(
            200,
            json={
                "payload": {"listens": page_two},
                "links": {
                    "next": "/1/user/alice/listens?next=cursor-2&payload=payload-2"
                },
            },
        ),
    ]

    since = date(2023, 1, 1)
    async with httpx.AsyncClient() as client:
        service = _ConcreteListenBrainzClient(client)
        listens = await service.fetch_listens("alice", since, token="secret", limit=3)

    assert listens == page_one + page_two[:1]
    assert route.call_count == 2

    min_ts = int(datetime.combine(since, datetime.min.time(), tzinfo=UTC).timestamp())
    first_query = dict(route.calls[0].request.url.params)
    second_query = dict(route.calls[1].request.url.params)

    assert first_query["count"] == "3"
    assert first_query["min_ts"] == str(min_ts)
    assert second_query["next"] == "cursor-1"
    assert second_query["payload"] == "payload-1"
    assert second_query["count"] == "1"
    assert second_query["min_ts"] == str(min_ts)
    assert route.calls[0].request.headers["Authorization"] == "Token secret"


@pytest.mark.unit
@pytest.mark.asyncio
async def test_fetch_listens_follows_string_cursor(respx_mock):
    route = respx_mock.get("https://api.listenbrainz.org/1/user/bob/listens")
    route.side_effect = [
        httpx.Response(
            200,
            json={
                "payload": {"listens": [{"idx": 1}]},
                "links": {"next": "/1/user/bob/listens?next=abc&payload=def"},
            },
        ),
        httpx.Response(
            200,
            json={
                "listens": [{"idx": 2}],
                "next": "?next=ghi&payload=jkl",
            },
        ),
        httpx.Response(200, json={"payload": {"listens": [{"idx": 3}]}}),
    ]

    async with httpx.AsyncClient() as client:
        service = _ConcreteListenBrainzClient(client)
        listens = await service.fetch_listens("bob", None, limit=5)

    assert listens == [{"idx": 1}, {"idx": 2}, {"idx": 3}]
    assert route.call_count == 3

    first_query = dict(route.calls[0].request.url.params)
    second_query = dict(route.calls[1].request.url.params)
    third_query = dict(route.calls[2].request.url.params)

    assert first_query["count"] == "5"
    assert second_query["next"] == "abc"
    assert second_query["payload"] == "def"
    assert second_query["count"] == "4"
    assert third_query["next"] == "ghi"
    assert third_query["payload"] == "jkl"
    assert third_query["count"] == "3"
