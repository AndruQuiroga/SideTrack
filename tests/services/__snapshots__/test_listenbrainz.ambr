# serializer version: 1
# name: test_get_cf_recommendations_contract[recommendations]
  list([
    dict({
      'artist_name': 'Artist A',
      'recording_mbid': 'rec1',
      'recording_name': 'Track A',
      'score': 0.9,
    }),
    dict({
      'artist_name': 'Artist B',
      'recording_mbid': 'rec2',
      'recording_name': 'Track B',
      'score': 0.8,
    }),
  ])
# ---
# name: test_get_cf_recommendations_contract[recordings]
  list([
    dict({
      'artist_name': 'Artist A',
      'recording_mbid': 'rec1',
      'recording_name': 'Track A',
      'score': 0.9,
    }),
    dict({
      'artist_name': 'Artist B',
      'recording_mbid': 'rec2',
      'recording_name': 'Track B',
      'score': 0.8,
    }),
  ])
# ---
# name: test_get_cf_recommendations_contract[recommended_recordings]
  list([
    dict({
      'artist_name': 'Artist A',
      'recording_mbid': 'rec1',
      'recording_name': 'Track A',
      'score': 0.9,
    }),
    dict({
      'artist_name': 'Artist B',
      'recording_mbid': 'rec2',
      'recording_name': 'Track B',
      'score': 0.8,
    }),
  ])
# ---
# name: test_listenbrainz_candidates_normalization
  list([
    dict({
      'artist': 'Artist A',
      'isrc': None,
      'recording_mbid': 'rec1',
      'score_cf': 0.9,
      'seeds': dict({
        'user': list([
          'alice',
        ]),
      }),
      'source': 'listenbrainz',
      'spotify_id': None,
      'title': 'Track A',
    }),
    dict({
      'artist': 'Artist B',
      'isrc': None,
      'recording_mbid': 'rec2',
      'score_cf': 0.8,
      'seeds': dict({
        'user': list([
          'alice',
        ]),
      }),
      'source': 'listenbrainz',
      'spotify_id': None,
      'title': 'Track B',
    }),
  ])
# ---
